### PR TITLE
Allow optional enforcing of implicit aliasing of tables (L011) and columns (L012)

### LINKED ISSUE #1400: Add configuration support for implicit table aliasing (Rule L011).
Hi 👋 

I'm working with a team that uses implicit aliasing as their standard practice - and wishes to remove `AS` from table references and joins.

To be able to fully implement SQLFluff as a linter, it would be great to add configuration support to L011 to make this possible.

Here's an example piece of code:

```
-- Snippet 1
SELECT
	a.name
	, b.house
FROM user AS a
JOIN house AS b ON a.house_id = b.id
```

At the moment this would pass L011 checks because AS is explicit.

The code we would like is:

```
-- Snippet 2
SELECT
	a.name
	, b.house
FROM user a
JOIN house b ON a.house_id = b.id
```

In our configuration, we would have to ignore Rule L011 - but we would actually like to use Rule L011 to fix `Snippet 1` in the examples above.

For example

```
-- Snippet 1
SELECT
	a.name
	, b.house
FROM user AS a
JOIN house AS b ON a.house_id = b.id

$ sqlfluff fix Snippet 1

>

-- Snippet 2
SELECT
	a.name
	, b.house
FROM user a
JOIN house b ON a.house_id = b.id
```

It would be great to have a configuration option - for example:

`aliasing: ["implicit", "explicit"]`

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made
Fixes #1400 

### Are there any other side effects of this change that we should be aware of?
Affects rules L011 and L012 which inherits from this. Default is left as explicit.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/rules/config_info.py [hunk 1]
```diff
@@ -94,6 +94,12 @@
         "validation": ["before", "after"],
         "definition": ("Should operator be placed before or after newlines."),
     },
+    "aliasing": {
+        "validation": ["implicit", "explicit"],
+        "definition": (
+            "Should alias have an explict AS or is implicit aliasing allowed?"
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/config_info.py:100]:
```suggestion
            "Should alias have an explict AS or is implicit aliasing required?"
```

File: test/fixtures/rules/std_rule_cases/L011.yml [hunk 2]
```diff
@@ -4,3 +4,48 @@ test_fail_implicit_alias:
...
+  fix_str: select foo.bar from (select 1 as bar) foo
+  configs:
+    rules:
+      L011:
+        aliasing: implicit
+
+test_fail_implicit_alias_implicit_multiple:
+  # Test implicit with multiple tables
+  fail_str: select foo.bar from (select 1 as bar) AS bar, (select 1 as foo) AS foo
+  fix_str: select foo.bar from (select 1 as bar) bar, (select 1 as foo) foo
```
Comment (barrywhart) [inline test/fixtures/rules/std_rule_cases/L011.yml:34]: Should it clean up the nested `SELECT`, also?
```suggestion
  fix_str: select foo.bar from (select 1 bar) bar, (select 1 foo) foo
```

File: test/fixtures/rules/std_rule_cases/L011.yml [hunk 3]
```diff
@@ -4,3 +4,48 @@ test_fail_implicit_alias:
...
+      L011:
+        aliasing: implicit
+
+test_fail_implicit_alias_implicit_newline:
+  # Add whitespace when fixing implicit aliasing
+  fail_str: |
+      select foo.bar from (select 1 as bar)
+      AS foo
+  fix_str: |
+      select foo.bar from (select 1 as bar)
```
Comment (barrywhart) [inline test/fixtures/rules/std_rule_cases/L011.yml:46]:
```suggestion
      select foo.bar from (select 1 bar)
```

File: test/fixtures/rules/std_rule_cases/L012.yml [hunk 4]
```diff
@@ -11,4 +11,29 @@ issue_561:
 
   configs:
     core:
-      dialect: snowflake
\ No newline at end of file
+      dialect: snowflake
+
```
Comment (barrywhart) [inline test/fixtures/rules/std_rule_cases/L012.yml:15]: Nitpick: Extra blank lines

File: test/fixtures/rules/std_rule_cases/L011.yml (hunk 2 above)
Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L011.yml:34]: No. That's rule L012 (also changed in this PR btw).

File: test/fixtures/rules/std_rule_cases/L011.yml (hunk 3 above)
Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L011.yml:46]: Same thing. This is rule 12, not rule 11. Resolving without taking the suggestion.

File: test/fixtures/rules/std_rule_cases/L012.yml (hunk 4 above)
Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L012.yml:14]: Fixed!

**Review by barrywhart — APPROVED** [review]
Looks good!

Heh, those "SELECT 1 AS foo" queries are confusing -- the last two words can easily be (mis-)interpreted as a column alias.

**Pull request merged**